pkg/authorization/authz.go

```diff
@@ -0,0 +1,130 @@
+// The response from this method may contains content that overrides the daemon response
+// This allows authZ plugins to filter privileged content
+//
+// If multiple authZ plugins are specified, the block/allow decision is based on ANDing all plugin results
+// For response manipulation, the response from each plugin is piped between plugins. Plugin execution order
+// is determined according to daemon parameters
+func NewCtx(authZPlugins []string, user, userAuthNMethod, requestMethod, requestUri string) *Ctx {
+	return &Ctx{authZPlugins: authZPlugins, user: user, userAuthNMethod: userAuthNMethod, requestMethod: requestMethod, requestUri: requestUri}
+}
+
+type Ctx struct {
+	user            string
+	userAuthNMethod string
+	requestMethod   string
+	requestUri      string
+	authZPlugins    []string
+	plugins         []*plugins.Plugin
+	rawRequest      []byte
+}
+
+// AuthRequest authorized the request to the docker daemon using authZ plugins
+func (a *Ctx) AuthZRequest(w http.ResponseWriter, r *http.Request) (err error) {
+
+	logrus.Debugf("AuthZ requset using plugins %s", a.authZPlugins)
+
+	for _, name := range a.authZPlugins {
+
+		p, err := plugins.Get(name, AuthZApiImplements)
+		if err != nil {
+			return err
+		}
+		a.plugins = append(a.plugins, p)
+	}
+
+	a.rawRequest, err = httputil.DumpRequest(r, true)
+	if err != nil {
+		return err
+	}
+
+	authReq := Request{
+		User:            a.user,
+		UserAuthNMethod: a.userAuthNMethod,
+		RequestMethod:   a.requestMethod,
+		RequestUri:      a.requestUri,
+		RawRequest:      a.rawRequest}
+
+	for _, plugin := range a.plugins {
+
+		authRes := &Response{}
+
+		err = plugin.Client.Call(AuthZApiRequest, authReq, authRes)
+
+		if err != nil {
+			return err
+		}
+
+		if !authRes.Allow {
+			return fmt.Errorf(authRes.Msg)
+		}
+	}
+
+	return nil
+}
+
+// AuthRequest authorized and manipulates the response from docker daemon using authZ plugins
+func (a *Ctx) AuthZResponse(version version.Version, rr *httptest.ResponseRecorder, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+
+	logrus.Debugf("AuthZ response using plugins %s", a.authZPlugins)
+
+	modifiedResponse := rr.Body.Bytes()
+	for _, plugin := range a.plugins {
+
+		authRes := &Response{}
+
+		authReq := Request{
+			User:            a.user,
+			UserAuthNMethod: a.userAuthNMethod,
+			RequestMethod:   a.requestMethod,
+			RequestUri:      a.requestUri,
+			RawRequest:      a.rawRequest,
+			RawResponse:     modifiedResponse}
+
+		err := plugin.Client.Call(AuthZApiResponse, authReq, authRes)
+
+		if err != nil {
+			return err
+		}
+
+		if !authRes.Allow {
+			return fmt.Errorf(authRes.Msg)
+		}
+
+		// Take latest response from authZ plugin
+		if authRes.ModifiedResponse != nil {
+			logrus.Debugf("Modify response")
+			modifiedResponse = authRes.ModifiedResponse
+		}
+	}
+
+	// Write headers and body
+	w.WriteHeader(rr.Code)
```

Comment:
Yes, based on the feedback we've modified the current api to enable both body, header and status code manipulation (see updated md or api.go).